packages/typegen/src/query/column-info.ts

```diff
@@ -232,21 +235,54 @@ export const defaultAnalysedQueryField = (f: QueryField): AnalysedQueryField =>
   column: undefined,
 })
 
-export const isFieldNotNull = (sql: string, field: QueryField) => {
+const nonNullableExpressionTypes = new Set([
+  'integer',
+  'numeric',
+  'string',
+  'boolean',
+  'list',
+  'array',
+  'keyword',
+  'parameter',
+  'constant',
+  'values',
+])
+export const isNonNullableField = (sql: string, field: QueryField) => {
   const ast = getHopefullyViewableAST(sql)
-  const getMatchingCountColumns = () =>
-    ast.type === 'select' &&
-    ast.columns &&
-    ast.columns.filter(c => {
-      if (c.expr.type !== 'call' || c.expr.function.name !== 'count') {
-        return false
-      }
-      const name = c.alias || 'count'
-      return field.name === name
-    })
-
-  const matchingCountColumns = getMatchingCountColumns()
-  return matchingCountColumns && matchingCountColumns.length === 1 // If we found exactly one field which looks like the result of a `count(...)`, we can be sure it's not null.
+  if (ast.type !== 'select' || ast.columns == null) {
+    return false
+  }
+  const nonNullableColumns = ast.columns.filter(c => {
+    if (c.expr.type !== 'call') {
+      return false
+    }
+    const name = c.alias?.name ?? c.expr.function.name
+    if (field.name !== name) {
+      return false
+    }
+    if (c.expr.function.name === 'count') {
+      // `count` is the only aggregation function, which never returns null.
+      return true
+    }
+    if (c.expr.function.name === 'coalesce') {
+      // let's try to check the args for nullability - starting at the end.
+      return c.expr.args
+        .slice(1)
+        .reverse()
+        .find(arg => {
```

Comment:
yes, it should, good find.

```suggestion
        .some(arg => {
```

This probably means we should add another test like `coalesce(sum(1), 1, null)`, which we'd expect to be nullable.

Reply:
Hmm not sure why it would make a difference. Given every `arg` should always be truthy, I'd have thought `.some` would return a boolean and `.find` would return a truthy object. The result is being tested for truthiness so the only reason I suggested the change was to make the intention clearer, not because it affects functionality. Having said that, that sounds like a sensible test to include either way.

Reply:
Well this change is actually not only cosmetic, it's also faster.

`find` will iterate through the whole array regardless, while `some` will stop at the first `true`-returning callback.

You might argue that the difference is negligible here and that might be so for most cases (typically we won't have 100s of args), but this is precisely the reason why I rarely rely on `truthy`, as explained above.

This is how apps get slow and subtle bugs creep in.

Reply:
`find` stops at the first truthy-returning callback too. You can see this in a node repl:

```
> [1,2,3,4,5,6,7].find(n => { console.log({ n }); return n > 3 })
{ n: 1 }
{ n: 2 }
{ n: 3 }
{ n: 4 }
4

> [1,2,3,4,5,6,7].some(n => { console.log({ n }); return n > 3 })
{ n: 1 }
{ n: 2 }
{ n: 3 }
{ n: 4 }
true
```

Note that neither goes through the whole array. And I still like truthiness checks (when TypeScript can prevent the gotchas). Either way, I like it well enough for cosmetic reasons so it doesn't really matter!

Reply:
yes, you are right again. where is my mind? :D

let's get this shipped.